mlir/lark/mlir.lark

```diff
@@ -154,66 +155,89 @@ attribute_dict : ("{" "}") | ("{" attribute_entry ("," attribute_entry)* "}")
 trailing_type     : ":" (function_type | function_result_type)
 
 // Operation results
-op_result         : ssa_id (":" integer_literal)?
+op_result         : ssa_id optional_int_literal
 op_result_list    : op_result ("," op_result)* "="
 
 // Trailing location (for debug information)
 location : string_literal ":" decimal_literal ":" decimal_literal
 trailing_location : ("loc" "(" location ")")
 
 // Undefined operations in all dialects
-generic_operation : string_literal "(" ssa_use_list? ")" attribute_dict? trailing_type
-custom_operation  : bare_id "." bare_id ssa_use_list? trailing_type
+generic_operation : string_literal "(" optional_ssa_use_list ")" optional_attr_dict trailing_type
+custom_operation  : bare_id "." bare_id optional_ssa_use_list trailing_type
 
 // Final operation definition
 // NOTE: "pymlir_dialect_ops" is defined externally by pyMLIR
-operation         : op_result_list? (pymlir_dialect_ops | custom_operation | generic_operation) trailing_location?
+operation         : optional_op_result_list (pymlir_dialect_ops | custom_operation | generic_operation | module | function) optional_trailing_loc
```

Comment:
where does `module`/`function` as an `operation` occur?

Reply:
Saw this in mlir langref.

> Operations can represent many different concepts, from higher-level concepts like function definitions, function calls, buffer allocations, view or slices of buffers, and process creation, to lower-level concepts like target-independent arithmetic, target-specific instructions, configuration registers, and logic gates.

From https://mlir.llvm.org/docs/LangRef/#high-level-structure

Reply:
OK